aten/src/ATen/native/SummaryOps.cpp

```diff
@@ -65,4 +134,112 @@ _bincount_cpu(const Tensor& self, const Tensor& weights, int64_t minlength) {
   });
 }
 
+
+std::tuple<Tensor,Tensor> _histogram_cpu(
+    const Tensor& self,
+    int64_t bins,
+    c10::optional<ArrayRef<double>> range,
+    const Tensor& weights,
+    bool density) {
+
+  // Weights having a different shape from input is not supported yet. TO DO:
+  // Add support for weights broadcastable to input
+  bool has_weights = weights.defined();
+  Tensor flattened_weights;
+  if (has_weights) {
+    TORCH_CHECK(
+        weights.sizes() == self.sizes(),
+        "histogram only supports input and weights of the same shape");
+    flattened_weights = weights.flatten(0).contiguous();
+  }
+
+
+  // If range is not defined, compute min and max from the values in the Tensor
+  Scalar min;
+  Scalar max;
+  if (range.has_value()) {
+    min = range.value()[0];
+    max = range.value()[1];
+  } else {
+    min = self.min().item();
+    max = self.max().item();
+  }
+
+  Tensor hist = AT_DISPATCH_ALL_TYPES(
+      self.scalar_type(), "histogram_cpu_uniformbins", [&] {
+        const auto scalar = weights.scalar_type();
+        if (scalar == ScalarType::Undefined || scalar == ScalarType::Float)
+          return _histogram_cpu_template_uniformbins<scalar_t, float>(
+              self.flatten(0).contiguous(),
+              bins,
+              flattened_weights,
+              min.to<scalar_t>(),
+              max.to<scalar_t>(),
+              density);
+        return _histogram_cpu_template_uniformbins<scalar_t, double>(
+            self.flatten(0).contiguous(),
+            bins,
+            flattened_weights.to(kDouble),
+            min.to<scalar_t>(),
+            max.to<scalar_t>(),
+            density);
+      });
+
+  if (density) { // Compute the density
+    double minval = min.to<double>();
+    double maxval = max.to<double>();
+    if (minval == maxval) {
+      minval -= 1;
+      maxval += 1;
+    }
+    hist = hist.to(kDouble);
+    hist *= bins / (maxval - minval) / hist.sum();
+  }
+
+  Tensor edges;
+  if (self.scalar_type() == kFloat) {
```

Comment:
Edges should be returned in the default scalar type if self is not double, and in double if self is double.

Reply:
That wouldn't be consistent with numpy, numpy will return edges in float scalar type if self is float. However, it does make sense to have it that way indeed.

Changing it to return edges in default scalar type if self is not floating type, and in self type if self type is floating-point.